api/workload/v2/workload.go

```diff
@@ -0,0 +1,219 @@
+	DefaultAgentAddress = "unix:///tmp/agent.sock"
+
+	_unixPathPrefix = "unix://"
+)
+
+// X509SVIDs is an X.509 SVID response from the SPIFFE Workload API.
+type X509SVIDs struct {
+	// SVIDs is a list of X509SVID messages, each of which includes a single
+	// SPIFFE Verifiable Identity Document, along with its private key and bundle.
+	SVIDs []*X509SVID
+
+	// CRL is a list of revoked certificates.
+	// Unimplemented.
+	CRL *pkix.CertificateList
+
+	// FederatedBundles are CA certificate bundles belonging to foreign Trust Domains
+	// that the workload should trust, keyed by the SPIFFE ID of the foreign domain.
+	// Unimplemented.
+	FederatedBundles map[string]*x509.CertPool
+}
+
+// Default returns the default SVID (the first in the list).
+//
+// See the SPIFFE Workload API standard Section 5.3
+// (https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Workload_API.md#53-default-identity)
+func (x *X509SVIDs) Default() *X509SVID {
+	return x.SVIDs[0]
+}
+
+// SVID is an X.509 SPIFFE Verifiable Identity Document.
+//
+// See https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md
+type X509SVID struct {
+	SPIFFEID     string
+	PrivateKey   crypto.Signer
+	Certificates []*x509.Certificate
+	TrustBundle  *x509.CertPool
+}
+
+// WorkloadX509SVIDWatcher is implemented by consumers who wish to be updated on SVID changes.
+type WorkloadX509SVIDWatcher interface {
+	// UpdateX509SVIDs indicates to the Watcher that the SVID has been updated
+	UpdateX509SVIDs(*X509SVIDs)
+
+	// OnError indicates an error occurred.
+	OnError(err error)
+
+	// OnConnection indicates a change in the connection state to the SPIFFE agent.
+	OnConnection(connected bool)
+}
+
+// X509SVIDClient interacts with the SPIFFE Workload API.
+type X509SVIDClient struct {
+	watcher  WorkloadX509SVIDWatcher
+	addr     string
+	wg       sync.WaitGroup
+	ctx      context.Context
+	cancelFn func()
+	backoff  *backoff
+}
+
+// Option configures the workload client.
+type Option func(*X509SVIDClient)
+
+// WithAddr specifies the unix socket address of the SPIFFE agent.
+func WithAddr(addr string) Option {
+	return func(w *X509SVIDClient) {
+		w.addr = addr
+	}
+}
+
+// NewX509SVIDClient returns a new Workload API client for X.509 SVIDs.
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		addr:     DefaultAgentAddress,
+		watcher:  watcher,
+		ctx:      ctx,
+		cancelFn: cancel,
+		backoff:  newBackoff(),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if !strings.HasPrefix(c.addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", c.addr)
+	}
+	return c, nil
+}
+
+// Start starts the client.
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
```

Comment:
Start() cannot be called more than once (results in two or more streams) and cannot be called after Stop() (since `run` will return immediately as the context will is cancelled). We should at a minimum document this.

Reply:
i've added docs and some state management around this